llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -8147,7 +8147,7 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
 
     if (!TLI.shouldExpandPartialReductionIntrinsic(cast<IntrinsicInst>(&I))) {
       setValue(&I,
-               DAG.getNode(ISD::PARTIAL_REDUCE_UADD, dl, AccVT, Acc, Input));
+               DAG.getNode(ISD::PARTIAL_REDUCE_UMLA, dl, AccVT, Acc, Input));
```

Comment:
Correct. For this initial building of the DAG I recommend ignoring the operand analysis and just always create the extra vector of ones operand.  Then you can implement a DAG combine to transform the `PARTIAL_REDUCE_UMLA` based on the operands (i.e. remove an explicit multiply and/or extensions).

Reply:
It looks like you've left this to `tryCombineToDotProduct`, but that means the node you create here is broken (since it's missing an operand, and is not meant to multiply anything).

Reply:
Done, made `MUL` instructions get generated all the time in the combine and expand functions. So if `Input2` is a splat vector of all 1s, the instruction is automatically removed. I have tested that this works by trying it with a splat vector of all 3s, and the `MUL` instructions are generated and remain in the output.